2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+# Introduction
+
+This folder contains the design for the charms.reactive 2.0 reboot.
+
+*Tip: Forget about hooks!*
+
+# Index
+
+[VISION.md](VISION.md) offers a short description of what charms.reactive is about and where we want to go.
+
+# In general
+
+Events trigger (set or unset) flags. Handlers react to flags and change them. Flags can have data attached to them.
+
+<!-- triggers are suggested here: https://github.com/juju-solutions/charms.reactive/issues/97 -->
+
+That's it.
+
+# Types of Events
+
+### The world around us has changed! [external]
+
+The external state outside of this unit has changed and we need to react to that.
+
+Ex: `relation.changed` and `config.changed`.
+
+### Lifecycle transition requested! [external]
+
+The operator has requested that we bring the application into a new state of its lifecycle.
+
+Ex: `install`, `start`, `stop`.
+
+<!-- hooks are just one of the many types of events, decoupling the reactive bus from hooks as proposed here: https://github.com/juju-solutions/charms.reactive/issues/90 -->
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
```

Comment:
@jamesbeedy Thanks for your feedback, no offence taken. We need strong discussion for something as important as this. I'm working out a bunch of examples of problems that are solved with this. See [the use-cases section](https://github.com/juju-solutions/charms.reactive/blob/2.0/2.0-proposal/readme.md#use-cases). Is that the format that you want? Let me know if you want me to write out a specific use-case.

To make the explanation a bit easier: inactive handlers are handlers that have run, whose preconditions are still true, but the flags haven't changed since the last run.

The current implementation doesn't rerun inactive handlers in the same hook but reruns them in the next hook. This isn't exactly "clean", especially for users who came into Juju post-reactive. A clean solution is to either always rerun inactive handlers or never rerun inactive handlers. I don't think either solution will limit the user since he can always mimic the other behavior with states and triggers.

> then they wouldn't be preconditions anymore

I agree that this term isn't ideal, we should find a better term. Calling them preconditions makes more sense for states, not for flags. Preconditions have two uses.

1. Handlers **react** to flags. The flags signal `something` and a handler reacts to that `something`. The preconditions state what that `something` is. With this explanation, it makes sense to say "when `something` hasn't changed, there is no need for the handler to react to it again". 
2. Preconditions are a way to cut down on the amount of code needed to find out if it is ok to do something. You should only change the MongoDB config files after MongoDB is installed.A handler having `mongodb.installed` in its preconditions doesn't have to manually check if MongoDB is really installed. It can assume that that is the case. This is still the case even if you don't rerun inactive handlers.